### ISSUE TITLE
afl-gcc doesn't instrument anymore

### ISSUE THREAD

**Issue by elboulangero** [AFLplusplus/AFLplusplus#2058]
Hello,

After updating the package alfplusplus in Debian, one of the tests doesn't work anymore. It seems that something broke in `afl-gcc`.

Here's a trace for **working** afl-gcc (version `4.09c-1`):

```
# export AFL_DEBUG=1

# afl-gcc     test-instr.c   -o test-instr
[D] DEBUG: Trying /usr/bin/../lib/afl/SanitizerCoverageLTO.so
[D] DEBUG: Trying /usr/bin/../lib/afl/cmplog-routines-pass.so
[D] DEBUG: Trying /usr/bin/../lib/afl/afl-gcc-pass.so
afl-cc++4.09c by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: GCC-GCC
[!] WARNING: You are using outdated instrumentation, install LLVM and/or gcc-plugin and use afl-clang-fast/afl-clang-lto/afl-gcc-fast instead!
[D] DEBUG: cd '/work/4.09c-1'; 'afl-gcc' 'test-instr.c' '-o' 'test-instr'
[D] DEBUG: Trying /usr/bin/../lib/afl/afl-compiler-rt.o
[D] DEBUG: rt=/usr/bin/../lib/afl/afl-compiler-rt.o obj_path=/usr/bin/../lib/afl/
[D] DEBUG: cd '/work/4.09c-1'; 'gcc' '-B' '/usr/bin/../lib/afl/' 'test-instr.c' '-o' 'test-instr' '-fPIC' '-g' '-O3' '-funroll-loops' '-D__AFL_COMPILER=1' '-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=1' '-D__AFL_COVERAGE()=int __afl_selective_coverage = 1;void __afl_coverage_discard();void __afl_coverage_skip();void __afl_coverage_on();void __afl_coverage_off();' '-D__AFL_COVERAGE_START_OFF()=int __afl_selective_coverage_start_off = 1;' '-D__AFL_COVERAGE_ON()=__afl_coverage_on()' '-D__AFL_COVERAGE_OFF()=__afl_coverage_off()' '-D__AFL_COVERAGE_DISCARD()=__afl_coverage_discard()' '-D__AFL_COVERAGE_SKIP()=__afl_coverage_skip()' '-D__AFL_FUZZ_TESTCASE_BUF=(__afl_fuzz_ptr ? __afl_fuzz_ptr : __afl_fuzz_alt_ptr)' '-D__AFL_FUZZ_TESTCASE_LEN=(__afl_fuzz_ptr ? *__afl_fuzz_len : (*__afl_fuzz_len = read(0, __afl_fuzz_alt_ptr, 1048576)) == 0xffffffff ? 0 : *__afl_fuzz_len)'
afl-as++4.09c by Michal Zalewski
[+] Instrumented 19 locations (64-bit, non-hardened mode, ratio 100%).
```

And now is **non-working** afl-gcc (version `4.20c-1`):

```
# export AFL_DEBUG=1

# afl-gcc     test-instr.c   -o test-instr
[D] DEBUG: Trying /usr/bin/../lib/afl/SanitizerCoveragePCGUARD.so
[D] DEBUG: Trying /usr/bin/../lib/afl/SanitizerCoverageLTO.so
[D] DEBUG: Trying /usr/bin/../lib/afl/cmplog-routines-pass.so
[D] DEBUG: Trying /usr/bin/../lib/afl/afl-gcc-pass.so
[D] DEBUG: Trying /usr/bin/../lib/afl/afl-compiler-rt.o
[D] DEBUG: rt=/usr/bin/../lib/afl/afl-compiler-rt.o
afl-cc++4.20c by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: GCC-GCC
[!] WARNING: You are using outdated instrumentation, install LLVM and/or gcc-plugin and use afl-clang-fast/afl-clang-lto/afl-gcc-fast instead!
[D] DEBUG: cd '/work/4.20c-1'; 'afl-gcc' 'test-instr.c' '-o' 'test-instr'
[D] DEBUG: cd '/work/4.20c-1'; 'gcc' '-B' '/usr/bin' 'test-instr.c' '-o' 'test-instr' '-fPIC' '-g' '-O3' '-funroll-loops' '-D__AFL_COMPILER=1' '-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=1' '-D__AFL_COVERAGE()=int __afl_selective_coverage = 1;void __afl_coverage_discard();void __afl_coverage_skip();void __afl_coverage_on();void __afl_coverage_off();' '-D__AFL_COVERAGE_START_OFF()=int __afl_selective_coverage_start_off = 1;' '-D__AFL_COVERAGE_ON()=__afl_coverage_on()' '-D__AFL_COVERAGE_OFF()=__afl_coverage_off()' '-D__AFL_COVERAGE_DISCARD()=__afl_coverage_discard()' '-D__AFL_COVERAGE_SKIP()=__afl_coverage_skip()' '(null)'
```

In this second trace, what's missing is the last line: `[+] Instrumented 19 locations (64-bit, non-hardened mode, ratio 100%).`

Comparing the 2 traces now...

First difference that stands out: for `4.09c-1`, the line `Trying /usr/bin/../lib/afl/afl-gcc-pass.so` triggers something (it's followed by a line `afl-cc++4.09c...`), while it doesn't trigger anything in version `4.20c-1`.... But, that's assuming logs come in order, is it true?

The actual issue is the value for the `-B` parameter given to gcc. For `4.09c-1` we have `'gcc' '-B' '/usr/bin/../lib/afl/'` while for `4.20c-1` we have `'gcc' '-B' '/usr/bin'`.. That's the problem.

By running the gcc line manualy, and fixing this value, it seems that I can get it to work:

```
# 'gcc' '-B' '/usr/bin/../lib/afl/' 'test-instr.c' '-o' 'test-instr' '-fPIC' '-g' '-O3' '-funroll-loops' '-D__AFL_COMPILER=1' '-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=1' '-D__AFL_COVERAGE()=int __afl_selective_coverage = 1;void __afl_coverage_discard();void __afl_coverage_skip();void __afl_coverage_on();void __afl_coverage_off();' '-D__AFL_COVERAGE_START_OFF()=int __afl_selective_coverage_start_off = 1;' '-D__AFL_COVERAGE_ON()=__afl_coverage_on()' '-D__AFL_COVERAGE_OFF()=__afl_coverage_off()' '-D__AFL_COVERAGE_DISCARD()=__afl_coverage_discard()' '-D__AFL_COVERAGE_SKIP()=__afl_coverage_skip()'
afl-as++4.20c by Michal Zalewski
[+] Instrumented 19 locations (64-bit, non-hardened mode, ratio 100%).
```

There's been some heavy refactoring of `src/afl-cc.c` between `4.09` and `4.20` so I suppose something broke.

I'm trying to find out what exactly, I'll follow up, but if someone can already give some hints at this point, that would be lovely!

Thanks!

**Comment by elboulangero** [comment]
I think I understand the issue. In the old version of the code, we had this:

https://github.com/AFLplusplus/AFLplusplus/blob/108fb0b29ad1586e668ba23e23a0eb1a13c45c49/src/afl-cc.c#L795-L798

The code used `-B <obj_path>`, that was set by previous call to `find_object()`. That was sloppy, as `find_object()` is called several times, and each time it updates `obj_path`, so it means we relied on the value of obj_path set by the last time find_object was called... Well, in most cases, `find_object` is called for object that belongs to aflplusplus (eg. `SanitizerCoverageLTO.so`), thus objects are located in `/usr/bin/../lib/afl/`, and so that's the value we had in `obj_path`.

Fast forward to today's code, we have this:

https://github.com/AFLplusplus/AFLplusplus/blob/1d17210d9fb0eb37ba866a3697643a9e4f37acd5/src/afl-cc.c#L2474-L2488

So the obj_path is derived from the location of `as`, and since the code in `find_object` looks into `/usr/bin/` first, of course it finds gcc `/usr/bin/as` first... Hence the bug.

I have the impression that the right fix here is simply to look for `afl-as` instead of `as` in the code above? What do you think?

**Comment by elboulangero** [comment]
Proposing https://github.com/AFLplusplus/AFLplusplus/pull/2059, I tested it, it fixes the issue:

```
# afl-gcc     test-instr.c   -o test-instr
[D] DEBUG: Trying /usr/bin/as
[D] DEBUG: Trying /usr/bin/SanitizerCoveragePCGUARD.so
[D] DEBUG: Trying /usr/bin/../lib/afl/SanitizerCoveragePCGUARD.so
[D] DEBUG: Trying /usr/bin/SanitizerCoverageLTO.so
[D] DEBUG: Trying /usr/bin/../lib/afl/SanitizerCoverageLTO.so
[D] DEBUG: Trying /usr/bin/cmplog-routines-pass.so
[D] DEBUG: Trying /usr/bin/../lib/afl/cmplog-routines-pass.so
[D] DEBUG: Trying /usr/bin/afl-gcc-pass.so
[D] DEBUG: Trying /usr/bin/../lib/afl/afl-gcc-pass.so
[D] DEBUG: Trying /usr/bin/afl-compiler-rt.o
[D] DEBUG: Trying /usr/bin/../lib/afl/afl-compiler-rt.o
[D] DEBUG: rt=/usr/bin/../lib/afl/afl-compiler-rt.o
afl-cc++4.20c by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: GCC-GCC
[!] WARNING: You are using outdated instrumentation, install LLVM and/or gcc-plugin and use afl-clang-fast/afl-clang-lto/afl-gcc-fast instead!
[D] DEBUG: cd '/work/4.20c-1'; 'afl-gcc' 'test-instr.c' '-o' 'test-instr'
[D] DEBUG: Trying /usr/bin/afl-as
[D] DEBUG: Trying /usr/bin/../lib/afl/afl-as
[D] DEBUG: cd '/work/4.20c-1'; 'gcc' '-B' '/usr/bin/../lib/afl' 'test-instr.c' '-o' 'test-instr' '-fPIC' '-g' '-O3' '-funroll-loops' '-D__AFL_COMPILER=1' '-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=1' '-D__AFL_COVERAGE()=int __afl_selective_coverage = 1;void __afl_coverage_discard();void __afl_coverage_skip();void __afl_coverage_on();void __afl_coverage_off();' '-D__AFL_COVERAGE_START_OFF()=int __afl_selective_coverage_start_off = 1;' '-D__AFL_COVERAGE_ON()=__afl_coverage_on()' '-D__AFL_COVERAGE_OFF()=__afl_coverage_off()' '-D__AFL_COVERAGE_DISCARD()=__afl_coverage_discard()' '-D__AFL_COVERAGE_SKIP()=__afl_coverage_skip()' '(null)'
afl-as++4.20c by Michal Zalewski
[+] Instrumented 19 locations (64-bit, non-hardened mode, ratio 100%).
```

**Comment by SonicStark** [comment]
Thank you for reporting this bug. I rewrote `find_object` before because it used to update `obj_path` in a fragile way as what you see.
And the bug was indeed due to my negligence, because  in most cases `afl-gcc` is installed in `/usr/local/bin` rather than `/usr/bin` where system-wide `as` installed. Your fix is right and I've left some comments in #2059.
Thank you again!

**Comment by elboulangero** [comment]
Fixed via #2059, can be closed
